### PR TITLE
Improve module compilation

### PR DESCRIPTION
**Pull request by Arthapz**

### CODE REVIEW

File: xmake/rules/c++/modules/modules_support/common.lua [hunk 1]
```diff
@@ -704,11 +708,32 @@ function install_module_target(target)
...
+    local cachedir = modules_cachedir(target)
+    local modulepath = module.path or module
+    local cached = localcache():get2("modules_paths", modulepath)
+    if cached then
+        if not os.exists(cached) then
+            os.mkdir(cached)
+        end
+        return cached
+    else
+        local hashed = hash.sha1(bytes(path.directory(modulepath) .. target:name())):sub(1, 6)
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/common.lua:731]:
```
local key = path.directory(modulepath) .. target:name()
local hashed = ash.uuid(key):split("-", {plain = true})[1]:lower()
```

File: xmake/rules/c++/modules/modules_support/clang.lua [hunk 2]
```diff
@@ -278,9 +273,14 @@ function generate_dependencies(target, sourcebatch, opt)
                     local compinst = target:compiler("cxx")
                     local compflags = compinst:compflags({sourcefile = file, target = target})
                     local flags = {}
+                    local next_flag = false
                     for _, flag in pairs(compflags) do
-                        if flag:startswith("-stdlib") or (flag:startswith("-f") and not flag:startswith("-fmodules")) or flag:startswith("-D") or flag:startswith("-U") or flag:startswith("-I") or flag:startswith("-isystem") then
+                        if flag == "-m64" or flag == "-g" or flag:startswith("-stdlib") or flag:startswith("-m") or (flag:startswith("-f") and not flag:startswith("-fmodule") and not flag:startswith("-fno-implicit-module-maps")) or flag:startswith("-D") or flag:startswith("-U") or flag:startswith("-I") or flag:startswith("-isystem") or next_flag then
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/clang.lua:278]: please use multiple lines.

```
                        if flag == "-m64" or flag == "-g" or flag:startswith("-stdlib") or flag:startswith("-m") or
                            (flag:startswith("-f") and not flag:startswith("-fmodule") and not flag:startswith("-fno-implicit-module-maps")) or
                            flag:startswith("-D") or flag:startswith("-U") or flag:startswith("-I") or flag:startswith("-isystem") or next_flag then
```

File: xmake/rules/c++/modules/modules_support/clang.lua [hunk 3]
```diff
@@ -278,9 +273,14 @@ function generate_dependencies(target, sourcebatch, opt)
                     local compinst = target:compiler("cxx")
                     local compflags = compinst:compflags({sourcefile = file, target = target})
                     local flags = {}
+                    local next_flag = false
                     for _, flag in pairs(compflags) do
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/clang.lua:277]: ipairs

File: xmake/rules/c++/modules/modules_support/common.lua [hunk 4]
```diff
@@ -297,24 +297,27 @@ function _parse_dependencies_data(target, moduleinfos)
...
+                        if not bmifile then
+                            local name = provide["logical-name"] .. bmi_extension(target)
+                            bmifile = path.join(get_outputdir(target,  name), name)
                         end
-                    else
-                        -- assume path with name
-                        local name = provide["logical-name"] .. bmi_extension(target)
-                        -- partition ":" character is invalid path character on windows
-                        -- @see https://github.com/xmake-io/xmake/issues/2954
-                        name = name:replace(":", "-")
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/common.lua:311]: why remove it?

File: xmake/rules/c++/modules/modules_support/common.lua [hunk 5]
```diff
@@ -297,24 +297,27 @@ function _parse_dependencies_data(target, moduleinfos)
...
+                        for _, output in ipairs(rule.outputs) do
+                            if output:endswith(".ifc") or output:endswith(".pcm") or output:endswith(".bmi") then
+                                bmifile = output
+                                break
+                            end
+                        end
+
+                        -- we didn't found the compiled module path, so we assume it
+                        if not bmifile then
+                            local name = provide["logical-name"] .. bmi_extension(target)
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/common.lua:312]: we still need fix issue #2954 here.

Comment (Arthapz) [reply xmake/rules/c++/modules/modules_support/common.lua:312]: yeah sorry, forgot about that

File: xmake/rules/c++/modules/modules_support/clang.lua [hunk 6]
```diff
@@ -277,7 +277,8 @@ function generate_dependencies(target, sourcebatch, opt)
                     for _, flag in ipairs(compflags) do
                         if flag == "-m64" or flag == "-g" or flag:startswith("-stdlib") or flag:startswith("-m") or
                            (flag:startswith("-f") and not flag:startswith("-fmodule") and not flag:startswith("-fno-implicit-module-maps")) or
-                           flag:startswith("-D") or flag:startswith("-U") or flag:startswith("-I") or flag:startswith("-isystem") or next_flag then
+                           flag:startswith("-D") or flag:startswith("-U") or flag:startswith("-I") or flag:startswith("-isystem") or next_flag or
+                           flag:startswith("-iframework") then
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/clang.lua:281]: `-iframework`?  `-framework`?

`-Ixxx`, `-I xxx`  
`-isystemxxx`, `-isystem xxx` 

are the correct flags

Comment (Arthapz) [reply xmake/rules/c++/modules/modules_support/clang.lua]: -iframework does exists, but i found a better way than enabling them one by one :D

File: xmake/rules/c++/modules/modules_support/gcc.lua [hunk 7]
```diff
@@ -205,14 +200,14 @@ function generate_dependencies(target, sourcebatch, opt)
...
                     local compinst = target:compiler("cxx")
                     local compflags = compinst:compflags({sourcefile = file, target = target})
-                    local flags = {}
-                    for _, flag in ipairs(compflags) do
-                        if flag:startswith("-std") or (flag:startswith("-f") and not flag:startswith("-fmodules")) or flag:startswith("-D") or flag:startswith("-U") or flag:startswith("-I") or flag:startswith("-isystem") then
-                            table.insert(flags, flag)
+                    for i, flag in ipairs(compflags) do
+                        -- exclude -fmodule* flags because, when they are set gcc try to find bmi of imported modules but they don't exists a this point of compilation
+                        if flag:startswith("-fmodule") then
+                            table.remove(compflags, i)
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/gcc.lua:206]: it's non-safe. compflags is being traversed. you should use a new copy.

or use `table.remove_if`

```
function test_remove_if(t)
    t:are_equal(table.remove_if({1, 2, 3, 4, 5, 6}, function (i, v) return (v % 2) == 0 end), {1, 3, 5})
    t:are_equal(table.remove_if({a = 1, b = 2, c = 3}, function (i, v) return (v % 2) == 0 end), {a = 1, c = 3})
end
```

**Pull request merged**